src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,157 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final int INIT_SIZE = 0;
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    MyLinkedList() {
+        size = INIT_SIZE;
+    }
+
     @Override
     public void add(T value) {
+        if (size != 0) {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+            size++;
+            return;
+        }
+        initFirstNode(value);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index != size) {
+            checkIndex(index);
+            if (index == 0) {
+                Node<T> newHead = new Node<>(null, value, head);
+                toLink(newHead, head);
+                head = newHead;
+            } else if (index > 0 && index < size) {
+                Node<T> indexNode = getNode(index);
+                Node<T> currentNode = new Node<>(null, value, null);
+                toLink(indexNode.prev, currentNode);
+                toLink(currentNode, indexNode);
+            }
+        }
+        add(value);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
+        checkIndex(index);
+        Node<T> indexNode = getNode(index);
+        if (indexNode != null) {
+            toLink(indexNode.prev, indexNode.next);
+            size--;
+            return indexNode.value;
+        }
```

Comment:
duplicated code(same in `remove(T object)` method), lets create some `unlink` method and move this part of the code there